_posts/2020-08-24-training-imagenet-with-r/training-imagenet-with-r.Rmd

```diff
@@ -0,0 +1,270 @@
+|   1  Tesla K80           Off  | 00000000:00:06.0 Off |                    0 |
+| N/A   76C    P0   114W / 149W |  10936MiB / 11441MiB |     26%      Default |
++-------------------------------+----------------------+----------------------+
+                                                                               
++-----------------------------------------------------------------------------+
+| Processes:                                                       GPU Memory |
+|  GPU       PID   Type   Process name                             Usage      |
+|=============================================================================|
++-----------------------------------------------------------------------------+
+```
+
+The `MirroredStrategy` can help us scale up to about 8 GPUs per compute instance; however, we are likely to need 16 instances with 8 GPUs each to train ImageNet in a reasonable time, see Jeremy Howard's post on [Training Imagenet in 18 Minutes](https://www.fast.ai/2018/08/10/fastai-diu-imagenet/). So where do we go from here?
+
+Welcome to `MultiWorkerMirroredStrategy`, this strategy can use not only multiple GPUs, but also multiple GPUs across multiple computers. To configure them, all we have to do is define a `TF_CONFIG` environment variable with the right addresses and run the exact same code in each compute instance.
+
+```{r eval=FALSE}
+library(tensorflow)
+
+partition <- 0
+Sys.setenv(TF_CONFIG = jsonlite::toJSON(list(
+    cluster = list(
+        worker = c("10.100.10.100:10090", "10.100.10.101:10090")
+    ),
+    task = list(type = 'worker', index = partition)
+), auto_unbox = TRUE))
+
+strategy <- tf$distribute$MultiWorkerMirroredStrategy(
+  cross_device_ops = tf$distribute$ReductionToOneDevice())
+
+alexnet::imagenet_partition(partition = partition) %>%
+  alexnet::alexnet_train(strategy = strategy, parallel = 6)
+```
+
+Please note that `partition` must change for each compute instance to uniquely identify it and the IP addresses need also to be adjusted. In addition, `data` should point to a different partition of ImageNet, which we can retrieve with `pins`; although, for convenience, `alexnet` contains similar code under `alexnet::imagenet_partition()`. Other than that, the code that you need to run in each compute instance is exactly the same.
+
+However, if we were use 16 machines with 8 GPUs each to train ImageNet, it would be quite time-consuming and error-prone to manually run code in each R session. So instead, we should think of making use of cluster-computing frameworks, like Apache Spark with [barrier execution](https://blog.rstudio.com/2020/01/29/sparklyr-1-1/#barrier-execution). If you are new to Spark, there are many resources available at [sparklyr.ai](https://sparklyr.ai). To learn just about running Spark and TensorFlow together, watch our [Deep Learning with Spark, TensorFlow and R](https://www.youtube.com/watch?v=Zm20P3ADa14) video.
```

Comment:
```suggestion
However, if we were to use 16 machines with 8 GPUs each to train ImageNet, it would be quite time-consuming and error-prone to manually run code in each R session. So instead, we should think of making use of cluster-computing frameworks, like Apache Spark with [barrier execution](https://blog.rstudio.com/2020/01/29/sparklyr-1-1/#barrier-execution). If you are new to Spark, there are many resources available at [sparklyr.ai](https://sparklyr.ai). To learn just about running Spark and TensorFlow together, watch our [Deep Learning with Spark, TensorFlow and R](https://www.youtube.com/watch?v=Zm20P3ADa14) video.
```

Reply:
Good eds. Also, "time consuming" only gets a hyphen if it's being used as an adjective [see here](https://english.stackexchange.com/questions/453229/time-consuming-vs-time-consuming).

Reply:
I think it's an adjective since it's describing how the code, but good to know!